### DISCUSSION TITLE
Adding Exchange and Queue Configuration on Kubernetes StatefulSet YAML file

### DISCUSSION THREAD

**Discussion by berbarroso** [docker-library/rabbitmq#616] [Q&A] [answered]
Hi,

May I ask if there is a configuration for adding Exchange and Queue for StatefulSet YAML file like "**_RABBITMQ_DEFAULT_VHOST_**" for Virtual host? I just want to add an exchange and queue using YAML file instead of manually adding it on RabbitMQ Management Site.

Thanks for answering and have a great day!

**Comment by lukebakken** [comment, marked as answer]
Use [definition import](https://www.rabbitmq.com/definitions.html#import-on-boot).

I also suggest using the [RabbitMQ Cluster Operator](https://www.rabbitmq.com/kubernetes/operator/operator-overview.html).

**Reply by berbarroso** [reply]
Hi,

Noted on this, I will try it.

**Reply by berbarroso** [reply]
**Hi, I tried putting it on ConfigMap and StatefulSet yaml files. I don't know if I input the codes correctly.**

**This is my source code for ConfigMap**

apiVersion: v1
kind: ConfigMap
metadata:
  name: rabbitmq-config
data:
  enabled_plugins: |
    [rabbitmq_federation,rabbitmq_management,rabbitmq_peer_discovery_k8s].
  rabbitmq_definitions.json: |
    {
      "rabbit_version": "3.11.9",
      "rabbitmq_version": "3.11.9",
      "product_name": "RabbitMQ",
      "product_version": "3.11.9",
      "users": [
        {
            "name": "hello-world",
            #"password": "12345",
            "password_hash": "JQ6+ZVMAIIpmGS/pXb9Q6elneY94TrchYGYJAKE9wtRiIpRt",
            "hashing_algorithm": "rabbit_password_hashing_sha256",
            "tags": "administrator",
            "limits": {}
        },
        {
            "name": "guest",
            #"password": "12345",
            "password_hash": "X5L0vwDQq2g8bu2Rr3oGc+uJiU+tRSFqSOj14w6zYqRK/lDU",
            "hashing_algorithm": "rabbit_password_hashing_sha256",
            "tags": "administrator",
            "limits": {}
        }
      ],
      "vhosts": [
        {
            "name": "hello-world"
        },
        {
            "name": "/"
        }
      ],
      "permissions": [
        {
            "user": "guest",
            "vhost": "hello-world",
            "configure": ".*",
            "write": ".*",
            "read": ".*"
        },
        {
            "user": "guest",
            "vhost": "/",
             "configure": ".*",
            "write": ".*",
            "read": ".*"
        },
        {
            "user": "hello-world",
            "vhost": "hello-world",
            "configure": ".*",
            "write": ".*",
            "read": ".*"
        }
      ],
      "topic_permissions": [],
      "parameters": [],
      "global_parameters": [
        {
            "name": "cluster_name",
            "value": "rabbit@73cf1fdf05d2"
        },
        {
            "name": "internal_cluster_id",
            "value": "rabbitmq-cluster-id-j-jeqGlk6rJYvqR_Tb06yw"
        }
      ],
      "policies": [],
      "queues": [
        {
            "name": "qq1",
             "vhost": "hello-world",
            "durable": true,
            "auto_delete": false,
            "arguments": {
            "x-queue-type": "quorum"
        }
        },
        {
            "name": "cq1",
            "vhost": "hello-world",
            "durable": true,
            "auto_delete": false,
            "arguments": {
            "x-queue-type": "classic"
        }
        }
      ],
      "exchanges": [
        {
            "name": "example",
            "vhost": "hello-world",
            "type": "fanout",
            "durable": true,
            "auto_delete": false,
            "internal": false,
            "arguments": {}
        }
      ],
      "bindings": [
        {
            "source": "example",
            "vhost": "hello-world",
            "destination": "qq1",
            "destination_type": "queue",
            "routing_key": "",
            "arguments": {}
        },
        {
            "source": "example",
            "vhost": "hello-world",
            "destination": "cq1",
            "destination_type": "queue",
            "routing_key": "1234",
            "arguments": {}
        }
      ]
    }

  rabbitmq.conf: |
    loopback_users.guest = false
    listeners.tcp.default = 5672
    management.load_definitions = /etc/rabbitmq/rabbitmq_definitions.json
    cluster_formation.peer_discovery_backend  = rabbit_peer_discovery_k8s
    cluster_formation.k8s.host = kubernetes.default.svc.cluster.local
    cluster_formation.k8s.address_type = hostname
    cluster_formation.node_cleanup.only_log_warning = true

**Reply by berbarroso** [reply]
**For StatefulSet, I don't know what to input on volume to mount the rabbitmq_definitions.json file on ConfigMap**

**This is my source code for StatefulSet**

apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: rabbitmq
  namespace: rabbits
spec:
  serviceName: rabbitmq
  replicas: 1
  #replicas: 4
  selector:
    matchLabels:
      app: rabbitmq
  template:
    metadata:
      labels:
        app: rabbitmq
    spec:
      serviceAccountName: rabbitmq
      initContainers: 
      - name: config
        image: busybox 
        command: ['/bin/sh', '-c', 'cp /tmp/config/rabbitmq.conf /config/rabbitmq.conf && ls -l /config/ && cp /tmp/config/enabled_plugins /etc/rabbitmq/enabled_plugins']
        volumeMounts:
        - name: config
          mountPath: /tmp/config/
          readOnly: false
        - name: config-file
          mountPath: /config/
        - name: plugins-file
          mountPath: /etc/rabbitmq/
      containers:
      - name: rabbitmq
        #image: rabbitmq:3.8-management
        image: rabbitmq:3.11.9-management
        ports:
        - containerPort: 4369
          name: discovery
        - containerPort: 5672
          name: amqp
        - containerPort: 15672
          name: management 
        env:
        - name: RABBIT_POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: RABBIT_POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: RABBITMQ_NODENAME
          value: rabbit@$(RABBIT_POD_NAME).rabbitmq.$(RABBIT_POD_NAMESPACE).svc.cluster.local
        - name: RABBITMQ_USE_LONGNAME 
          value: "true"
        - name: RABBITMQ_CONFIG_FILE
          value: "/config/rabbitmq"
        - name: RABBITMQ_ERLANG_COOKIE
          valueFrom:
            secretKeyRef:
              name: rabbit-secret
              key: RABBITMQ_ERLANG_COOKIE
        - name: K8S_HOSTNAME_SUFFIX
          value: .rabbitmq.$(RABBIT_POD_NAMESPACE).svc.cluster.local
        volumeMounts:
        - name: data
          mountPath: /var/lib/rabbitmq
          readOnly: false
        - name: config-file
          mountPath: /config/
        - name: plugins-file #for plugin config
          mountPath: /etc/rabbitmq/
      volumes: #for config
      - name: config-file
        emptyDir: {}
      - name: plugins-file
        emptyDir: {}
      - name: config
        configMap:
          name: rabbitmq-config
          defaultMode: 0755
  volumeClaimTemplates:
  - metadata:
      name: data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      #storageClassName: "standard"
      storageClassName: "oci" 
      resources:
        requests:
          storage: 50Mi
---
apiVersion: v1
kind: Service
metadata:
  name: rabbitmq
spec:
  type: LoadBalancer
  ports:
  - port: 4369
    targetPort: 4369
    name: discovery
  - port: 5672
    targetPort: 5672
    name: amqp
  - port: 15672
    name: management 
  selector:
    app: rabbitmq

**Reply by ChunyiLyu** [reply]
Hi @berbarroso you can find two import definitions examples in Cluster Operator. Hope it helps!
https://github.com/rabbitmq/cluster-operator/tree/main/docs/examples/import-definitions
https://github.com/rabbitmq/cluster-operator/tree/main/docs/examples/import-https-url-definition

**Reply by berbarroso** [reply]
I will try these examples. Thanks!

**Reply by berbarroso** [reply]
Hi, 

I already tried the definitions examples and it's working, thank you!

I have another question, on the definitions.json file, how to decode the password key for  "password_hash"?

Thanks for answering! :)

![image](https://user-images.githubusercontent.com/78010659/221131827-f05a5e12-b87f-4be2-8f63-af7a3384dd80.png)

**Reply by lukebakken** [reply]
> how to decode the password key for "password_hash"?

You can't decode it, but you can _encode_ a password to use in that field. Here are several methods:

https://stackoverflow.com/questions/41306350/how-to-generate-password-hash-for-rabbitmq-management-http-api

We also just added a `rabbitmqctl` command in 3.11.9 to do that:

```
C:\Users\bakkenl\rmq-server\rabbitmq_server-3.11.9
> .\sbin\rabbitmqctl.bat hash_password foobarbaz
Will hash password foobarbaz
kYA8z4NaSdgh2M/ZdnumJUMif9X6Zmyh/cP77d9+V3fXNdbU
```

**Reply by berbarroso** [reply]
This is noted and thank you everyone for helping. :)
